File: typescript/packages/core/src/resources/retrieval.ts
/**
 * Retrieval resource — the agent-native three-step knowledge-graph surface
 * mounted at `/api/v1/query`:
 *
 * - `discover(sandboxId, …)`  → typed menu of relevant context candidates.
 * - `interpret(sandboxId, …)` → 1–2 paragraph brief for one candidate.
 * - `search(sandboxId, …)`    → deep matrix retrieval with a synthesized answer.
 *
 * All three accept either a `window` (a {@link WindowLike} — typically a
 * {@link ContextWindow} from `client.contextWindow.create()`) or a raw
 * `history` array of recent chat turns. When both are set `window` wins.
 * Server caps at 20 turns.
 */

import { BaseResource } from './base.js';
import type { WindowLike } from '../context-window/types.js';

export type ChatRole = 'user' | 'assistant' | 'system';

/**
 * Per-call cost telemetry attached to retrieval responses.
 *
 * `microcents` is the authoritative integer cost reported by the server.
 * `usd` is a display-only convenience (`microcents / 1_000_000`, rounded
 * to 6 decimals) — clients summing across responses MUST sum `microcents`
 * and divide once at display to avoid float-rounding drift.
 *
 * The field is optional on each response envelope: populated when the
 * server is configured to surface cost (`gate_mode` of `shadow` or
 * `enforce`) and absent / `null` when `gate_mode` is `off`.
 *
 * `gate_mode` indicates whether the server is enforcing credit balances
 * (`enforce`), reporting cost without enforcement (`shadow`), or not
 * tracking cost at all (`off`). Use it to detect "shadow mode is on —
 * cost is approximate" without inferring from a null `deduction_id`.
 */
export interface CostInfo {
  /**
   * Authoritative integer cost in USD microcents (1e-6 USD). Sum this
   * field — not `usd` — to aggregate cost across calls.
   */
  microcents: number;
  /**
   * Display-only USD convenience: `microcents / 1_000_000`, rounded to
   * 6 decimals. Float; do not sum for accounting.
   */
  usd?: number;
  /**
   * Opaque ledger identifier returned by the server; use it to join
   * against billing records you fetch from the server. May be `null`
   * (e.g. shadow mode where no ledger row was written, or paths that
   * skip settlement).
   */
  deduction_id?: string | null;
  /**
   * Server cost-tracking mode for this request. `off` means cost is
   * not tracked, `shadow` means cost is reported without enforcement,
   * `enforce` means credit balances are enforced.
   */
  gate_mode: 'off' | 'shadow' | 'enforce';
}

export interface ChatMessage {
  role: ChatRole;
  content: string;
  /**
   * Optional named participant for this turn. When set, adapters that
   * push a chat message through to ingestion forward this as the
   * envelope's `speaker` field, retiring the legacy `[author=…]`
   * content-prefix convention. Caller-decides the literal value
   * (`'Alice'`, `'support-bot'`, an email address, …); when absent,
   * adapters typically fall back to capitalizing `role`.
   */
  name?: string;
}

export interface DiscoveryItem {
  id: string;
  score: number;
  /**
   * One-sentence summary. Populated by `copass/copass_1.0` (path string);
   * empty under `copass/copass_2.0` — use `subgraph` instead.
   */
  summary: string;
  /**
   * Tuple of canonical IDs to pin /interpret retrieval to this item.
   *
   * - Under `copass/copass_1.0` this is the full hierarchical path
   *   (root → leaf).
   * - Under `copass/copass_2.0` this is the matched canonical plus
   *   every sub-graph node the query graph identified inside it.
   */
  canonical_ids: string[];
  /**
   * Pre-rendered ASCII tree of the matched canonical's sub-graph,
   * with matched nodes highlighted (⭐) and event timestamps inline.
   * Only populated by the `copass/copass_2.0` preset; absent under
   * `copass/copass_1.0` (which puts a path breadcrumb in `summary`).
   */
  subgraph?: string | null;
  /**
   * Names of the entities in the user's question that this item
   * satisfies. Helps the agent reason about which parts of the
   * question each item answers. Only populated by the
   * `copass/copass_2.0` preset.
   */
  matched_query_nodes?: string[] | null;
  /**
   * Distinct source file paths this item's canonicals were extracted
   * from, ordered by extraction frequency descending. Populated server-
   * side from `metadata_discovered(file_path=…)` events the ingestion
   * pipeline stamps onto each canonical when the source chunk carried
   * a `[File: …]` header (code ingest).
   *
   * Eliminates the `discover → get_origin` round-trip when the agent
   * only needs to know which files to open: read this field first;
   * fall back to {@link RetrievalResource.getOrigin} only when this
   * list is empty.
   *
   * Empty for legacy sandboxes that predate file-path stamping,
   * conversation-only ingests with no file markers, or items where
   * every backing canonical was deduped from a previously-unstamped
   * extraction.
   */
  file_paths?: string[];
}

export interface DiscoverRequest {
  query: string;
  /**
   * Preferred: a {@link WindowLike} (e.g. a {@link ContextWindow}). The
   * resource reads turns via `window.getTurns()` at call time and passes
   * them to the server as `history`.
   */
  window?: WindowLike;
  /** Raw chat turns. Used only when `window` is not supplied. */
  history?: ChatMessage[];
  project_id?: string;
  reference_date?: string;
  /**
   * Retrieval preset selecting the discovery shape. Defaults to
   * `copass/copass_1.0` server-side when omitted. Under
   * `copass/copass_2.0` items carry an additional `subgraph` field
   * with a pre-rendered ASCII tree of the matched canonical, plus a
   * `matched_query_nodes` list of the question entities that resolved
   * to it. The `:thinking` suffix is NOT accepted on `/discover`.
   */
  preset?: SearchPreset;
}

export interface DiscoverResponse {
  /** Markdown title + description orienting the caller on the response shape and how to use it. */
  header: string;
  items: DiscoveryItem[];
  count: number;
  sandbox_id: string;
  project_id?: string;
  query: string;
  /** Short actionable pointer for what to do after picking items. */
  next_steps: string;
  /**
   * Optional per-call cost telemetry. Populated by the server when
   * retrieval has a billable cost (`gate_mode` of `shadow` or
   * `enforce`); absent / `null` when `gate_mode` is `off` or when the
   * server omits the field. See {@link CostInfo} for field semantics.
   */
  cost?: CostInfo | null;
}

export interface InterpretRequest {
  query: string;
  /**
   * One or more tuples of canonical IDs to pin interpretation to.
   * Feed the `canonical_ids` list from each DiscoveryItem you want
   * to include.
   */
  items: string[][];
  /** Preferred: a {@link WindowLike}. Wins over `history` when both set. */
  window?: WindowLike;
  /** Raw chat turns. Used only when `window` is not supplied. */
  history?: ChatMessage[];
  project_id?: string;
  reference_date?: string;
  preset?: SearchPreset;
  /** Cap on brief length. Accepts 100–16000; omit for the server default. */
  max_tokens?: number;
}

export interface InterpretCitation {
  canonical_id: string;
  name: string;
  relevance: number;
}

export interface InterpretResponse {
  brief: string;
  citations: InterpretCitation[];
  /** Echo of the `items` tuples the caller sent, for correlation. */
  items: string[][];
  sandbox_id: string;
  project_id?: string;
  query: string;
  /**
   * Optional per-call cost telemetry. Populated by the server when
   * retrieval has a billable cost (`gate_mode` of `shadow` or
   * `enforce`); absent / `null` when `gate_mode` is `off` or when the
   * server omits the field. See {@link CostInfo} for field semantics.
   */
  cost?: CostInfo | null;
}

/**
 * Retrieval presets accepted by the Copass API.
 *
 * Canonical names (preferred):
 *   - `copass/copass_1.0` — path-discovery (low-latency default)
 *   - `copass/copass_2.0` — hierarchical-fused per-node embeddings
 *
 * Short aliases `copass/1.0` and `copass/2.0` are also accepted by the
 * server and resolve to the same SearchMatrix; new code should prefer
 * the canonical names.
 *
 * Append `:thinking` to any base preset (e.g. `copass/copass_2.0:thinking`)
 * to run an LLM pre-pass that decomposes the question into sub-questions,
 * executes the base preset on each, and synthesizes one combined answer.
 * The `:thinking` suffix is `/search`-only — `/interpret` and `/discover`
 * reject it.
 */
export type SearchPreset =
  // Canonical names
  | 'copass/copass_1.0'
  | 'copass/copass_2.0'
  | 'copass/copass_1.0:thinking'
  | 'copass/copass_2.0:thinking'
  // Short aliases (kept for backward-compat)
  | 'copass/1.0'
  | 'copass/2.0'
  | 'copass/1.0:thinking'
  | 'copass/2.0:thinking';

export interface SearchRequest {
  query: string;
  /** Preferred: a {@link WindowLike}. Wins over `history` when both set. */
  window?: WindowLike;
  /** Raw chat turns. Used only when `window` is not supplied. */
  history?: ChatMessage[];
  project_id?: string;
  reference_date?: string;
  preset?: SearchPreset;
  detail_level?: 'concise' | 'detailed';
  max_tokens?: number;
}

export interface SearchResponse {
  answer: string;
  preset: SearchPreset;
  execution_time_ms: number;
  warnings?: string[];
  sandbox_id: string;
  project_id?: string;
  query: string;
  /**
   * Optional per-call cost telemetry. Populated by the server when
   * retrieval has a billable cost (`gate_mode` of `shadow` or
   * `enforce`); absent / `null` when `gate_mode` is `off` or when the
   * server omits the field. See {@link CostInfo} for field semantics.
   */
  cost?: CostInfo | null;
}

/**
 * Extract `history` from a request: `window.getTurns()` takes precedence,
 * then the explicit `history` array, then an empty array. Strips `window`
 * from the body so it doesn't get serialized.
 */
function resolveBody<T extends { window?: WindowLike; history?: ChatMessage[] }>(
  request: T,
): Omit<T, 'window'> & { history: ChatMessage[] } {
  const { window, history, ...rest } = request;
  const resolved = window ? window.getTurns() : history ?? [];
  return { ...rest, history: resolved } as Omit<T, 'window'> & { history: ChatMessage[] };
}

/**
 * Request body for `client.retrieval.getOrigin(sandboxId, …)` — entity →
 * source file lookup. Pair with `discover`: after the caller picks items
 * from the menu, `getOrigin` returns the files those canonicals were
 * extracted from so an agent can localize its next action.
 */
export interface OriginsRequest {
  /**
   * Canonical IDs to look up. Typically the `canonical_ids` arrays from
   * `DiscoveryItem`s the caller chose. Capped at 100 per request server-side.
   */
  canonical_ids: string[];
  /**
   * Per-canonical cap on returned files (1–50). Files are sorted by how
   * many extractions of the canonical came from each, descending.
   * Defaults to 10 server-side when omitted.
   */
  limit_per_canonical?: number;
}

/** One source file recorded against a canonical. */
export interface OriginFile {
  /**
   * Source file path as stamped at ingest time. Typically a repo-relative
   * POSIX path (e.g. `src/foo/bar.py`). The literal value the ingestion
   * pipeline recorded — caller decides how to resolve it.
   */
  file_path: string;
  /**
   * Number of independent extractions of this canonical that came from
   * this file. Higher = the canonical's center-of-mass is here.
   */
  extraction_count: number;
}

/** Per-canonical roll-up of source files. */
export interface OriginEntry {
  /** Echo of the canonical_id the caller supplied. */
  canonical_id: string;
  /**
   * Distinct files the canonical was extracted from, ordered by
   * `extraction_count` descending. Empty when the canonical has no
   * recorded file-path metadata (e.g. sandboxes ingested before
   * file-path stamping shipped).
   */
  files: OriginFile[];
}

export interface OriginsResponse {
  sandbox_id: string;
  /**
   * One entry per requested `canonical_id`, in the same order. Entries
   * for canonicals with no recorded files come back with `files=[]`, so
   * the response stays positionally aligned with the input list.
   */
  origins: OriginEntry[];
  /**
   * Optional per-call cost telemetry — populated under the same
   * `gate_mode` rules as the other retrieval responses. See {@link CostInfo}.
   */
  cost?: CostInfo | null;
}

export class RetrievalResource extends BaseResource {
  discover(sandboxId: string, request: DiscoverRequest): Promise<DiscoverResponse> {
    return this.post<DiscoverResponse>(
      `/api/v1/query/sandboxes/${encodeURIComponent(sandboxId)}/discover`,
      resolveBody(request),
    );
  }

  interpret(sandboxId: string, request: InterpretRequest): Promise<InterpretResponse> {
    return this.post<InterpretResponse>(
      `/api/v1/query/sandboxes/${encodeURIComponent(sandboxId)}/interpret`,
      resolveBody(request),
    );
  }

  search(sandboxId: string, request: SearchRequest): Promise<SearchResponse> {
    return this.post<SearchResponse>(
      `/api/v1/query/sandboxes/${encodeURIComponent(sandboxId)}/search`,
      resolveBody(request),
    );
  }

  /**
   * Look up source files for one or more canonical entities. Use this
   * after `discover` to localize the agent's next action — e.g. open
   * the file with a native read tool. Cheaper than `search`: no LLM
   * legs, no synthesis, just an indexed DB lookup.
   */
  getOrigin(sandboxId: string, request: OriginsRequest): Promise<OriginsResponse> {
    return this.post<OriginsResponse>(
      `/api/v1/query/sandboxes/${encodeURIComponent(sandboxId)}/origins`,
      request,
    );
  }
}

File: python/copass-core/src/copass_core/resources/retrieval.py
"""Retrieval resource — ``discover`` / ``interpret`` / ``search``.

Hand-ported from ``typescript/packages/core/src/resources/retrieval.ts``.

All three accept either a ``window`` (a :class:`WindowLike` — typically
a ``ContextWindow``) or a raw ``history`` list. When ``window`` is set
it wins and ``history`` is ignored. Server caps at 20 turns.

All three responses MAY carry an optional ``cost`` sub-object. When
the deployment is configured with ``gate_mode == "off"``, ``cost`` is
absent / ``None``; in ``"shadow"`` and ``"enforce"`` modes it carries
``{microcents, usd, deduction_id, gate_mode}``. The Python type for
that sub-object is :class:`copass_core.types.CostInfo`; consumers that
want it parsed call ``CostInfo.from_dict(response["cost"])``.

TODO: replace the ``Dict[str, Any]`` returns with typed
``DiscoverResponse`` / ``InterpretResponse`` / ``SearchResponse``
envelopes so ``cost`` (and the rest of the envelope) gets IDE
autocomplete. Held back from this change because the broader retrieval
envelope (header / items / brief / citations / answer / preset /
execution_time_ms / etc.) is not yet typed anywhere in this SDK and
typing only one nested field is inconsistent. Tracked separately.
"""

from __future__ import annotations

from dataclasses import asdict
from typing import Any, Dict, List, Optional

from copass_core.resources.base import BaseResource
from copass_core.types import ChatMessage, SearchPreset, WindowLike


def _turns_from(window: Optional[WindowLike]) -> List[Dict[str, str]]:
    if window is None:
        return []
    turns = window.get_turns()
    if not turns:
        return []
    return [_turn_to_dict(t) for t in turns]


def _turn_to_dict(turn: Any) -> Dict[str, str]:
    """Accept either a :class:`ChatMessage` dataclass or a plain dict
    (``{"role": ..., "content": ...}``) so callers can mix and match."""
    if isinstance(turn, ChatMessage):
        return {"role": turn.role, "content": turn.content}
    if isinstance(turn, dict) and "role" in turn and "content" in turn:
        return {"role": str(turn["role"]), "content": str(turn["content"])}
    raise ValueError(
        f"Unexpected turn type {type(turn).__name__}; expected ChatMessage or dict."
    )


def _history(
    window: Optional[WindowLike],
    history: Optional[List[Any]],
) -> List[Dict[str, str]]:
    """``window`` wins; fall back to ``history``; empty list if both None."""
    if window is not None:
        return _turns_from(window)
    if history:
        return [_turn_to_dict(t) for t in history]
    return []


class RetrievalResource(BaseResource):
    """``/api/v1/query/sandboxes/{sandbox_id}/{discover,interpret,search}``."""

    async def discover(
        self,
        sandbox_id: str,
        *,
        query: str,
        window: Optional[WindowLike] = None,
        history: Optional[List[Any]] = None,
        project_id: Optional[str] = None,
        reference_date: Optional[str] = None,
        preset: Optional[SearchPreset] = None,
    ) -> Dict[str, Any]:
        """Ranked menu of context items relevant to ``query``.

        Window-aware: repeated calls skip items already surfaced
        earlier in the conversation.

        ``preset`` selects the discovery shape (defaults to
        ``copass/copass_1.0`` server-side when omitted). Under
        ``copass/copass_2.0`` each returned item carries an additional
        ``subgraph`` field with a pre-rendered ASCII tree of the matched
        canonical's sub-graph, plus a ``matched_query_nodes`` list of
        the question entities that resolved to it. The ``:thinking``
        suffix is NOT accepted on ``discover``.
        """
        body: Dict[str, Any] = {
            "query": query,
            "history": _history(window, history),
        }
        if project_id is not None:
            body["project_id"] = project_id
        if reference_date is not None:
            body["reference_date"] = reference_date
        if preset is not None:
            body["preset"] = preset
        return await self._post(
            f"/api/v1/query/sandboxes/{sandbox_id}/discover",
            body,
        )

    async def interpret(
        self,
        sandbox_id: str,
        *,
        query: str,
        items: List[List[str]],
        window: Optional[WindowLike] = None,
        history: Optional[List[Any]] = None,
        project_id: Optional[str] = None,
        reference_date: Optional[str] = None,
        preset: Optional[SearchPreset] = None,
        max_tokens: Optional[int] = None,
    ) -> Dict[str, Any]:
        """Synthesized 1–2 paragraph brief pinned to specific
        ``items`` picked from a prior ``discover`` call."""
        body: Dict[str, Any] = {
            "query": query,
            "items": items,
            "history": _history(window, history),
        }
        if project_id is not None:
            body["project_id"] = project_id
        if reference_date is not None:
            body["reference_date"] = reference_date
        if preset is not None:
            body["preset"] = preset
        if max_tokens is not None:
            body["max_tokens"] = max_tokens
        return await self._post(
            f"/api/v1/query/sandboxes/{sandbox_id}/interpret",
            body,
        )

    async def search(
        self,
        sandbox_id: str,
        *,
        query: str,
        window: Optional[WindowLike] = None,
        history: Optional[List[Any]] = None,
        project_id: Optional[str] = None,
        reference_date: Optional[str] = None,
        preset: Optional[SearchPreset] = None,
        detail_level: Optional[str] = None,
        max_tokens: Optional[int] = None,
    ) -> Dict[str, Any]:
        """One-shot synthesized natural-language answer."""
        body: Dict[str, Any] = {
            "query": query,
            "history": _history(window, history),
        }
        if project_id is not None:
            body["project_id"] = project_id
        if reference_date is not None:
            body["reference_date"] = reference_date
        if preset is not None:
            body["preset"] = preset
        if detail_level is not None:
            body["detail_level"] = detail_level
        if max_tokens is not None:
            body["max_tokens"] = max_tokens
        return await self._post(
            f"/api/v1/query/sandboxes/{sandbox_id}/search",
            body,
        )

    async def get_origin(
        self,
        sandbox_id: str,
        *,
        canonical_ids: List[str],
        limit_per_canonical: Optional[int] = None,
    ) -> Dict[str, Any]:
        """Look up source files for one or more canonical entities.

        Pair with ``discover``: after the caller picks items from the
        menu, ``get_origin`` returns the files those canonicals were
        extracted from so an agent can localize its next action.

        ``canonical_ids`` is at least 1, at most 100 per call.
        ``limit_per_canonical`` (1–50) caps the returned file count per
        canonical; omit to take the server default.

        The response shape is::

            {
              "sandbox_id": "...",
              "origins": [
                {
                  "canonical_id": "...",
                  "files": [
                    {"file_path": "src/foo.py", "extraction_count": 3},
                    ...
                  ]
                },
                ...
              ],
              "cost": {...} | None
            }

        Canonicals with no recorded file metadata come back with
        ``files=[]``, so the ``origins`` list is positionally aligned
        with ``canonical_ids``.
        """
        body: Dict[str, Any] = {"canonical_ids": canonical_ids}
        if limit_per_canonical is not None:
            body["limit_per_canonical"] = limit_per_canonical
        return await self._post(
            f"/api/v1/query/sandboxes/{sandbox_id}/origins",
            body,
        )


__all__ = ["RetrievalResource"]
